File: Source/Game/FirstPersonCameraManager.cs
using System.Collections.Generic;
using WCSharp.Api;
using static WCSharp.Api.Common;

public class FirstPersonCamera
{
    private const float timerPeriod = 0.001f;
    private const float FIRSTPERSON_ANGLE_PER_PERIOD = 0.3f;

    private timer forceCamTimer;
    private unit hero;
    private player player;
    private Dictionary<string, bool> keyDownState;
    private string lastUnitAnimation;

    public FirstPersonCamera(unit hero, player player)
    {
        this.hero = hero;
        this.player = player;
        this.keyDownState = new Dictionary<string, bool>
        {
            { "UP", false },
            { "DOWN", false },
            { "LEFT", false },
            { "RIGHT", false }
        };
    }

    public bool IsFirstPerson() => forceCamTimer != null;

    public void ToggleFirstPerson(bool active)
    {
        if (active)
        {
            if (forceCamTimer == null)
            {
                forceCamTimer = CreateTimer();
                TimerStart(forceCamTimer, timerPeriod, true, UpdateCamera);
            }
        }
        else
        {
            if (forceCamTimer != null)
            {
                DestroyTimer(forceCamTimer);
                forceCamTimer = null;
                ResetCamera();
            }
        }
    }

    private void UpdateCamera()
    {
        if (hero == null || !hero.Alive) return;

        float fwd = 0;

        if (!(keyDownState["LEFT"] && keyDownState["RIGHT"]) && (keyDownState["LEFT"] || keyDownState["RIGHT"]))
        {
            float angle = GetUnitFacing(hero);
            if (keyDownState["LEFT"]) angle += FIRSTPERSON_ANGLE_PER_PERIOD;
            if (keyDownState["RIGHT"]) angle -= FIRSTPERSON_ANGLE_PER_PERIOD;
            BlzSetUnitFacingEx(hero, angle);
        }

        if (!(keyDownState["UP"] && keyDownState["DOWN"]) && (keyDownState["UP"] || keyDownState["DOWN"]))
        {
            float moveSpeed = GetUnitMoveSpeed(hero);
            float movePerTick = moveSpeed * timerPeriod;

            Kitty kitty = Globals.ALL_KITTIES[player];

            if (kitty.Slider.IsEnabled())
            {
                movePerTick = 0.2f;
            }

            float angle = Rad2Deg(GetUnitFacing(hero));
            if (keyDownState["UP"]) fwd += movePerTick;
            if (keyDownState["DOWN"]) fwd -= movePerTick;

            float oldX = GetUnitX(hero);
            float oldY = GetUnitY(hero);

            float newX = oldX + (fwd * Cos(angle));
            float newY = oldY + (fwd * Sin(angle));

            if (!Globals.GAME_ACTIVE && Regions.BarrierRegion.Contains(newX, newY))
            {
                return;
            }

            if (IsTerrainPathable(newX, oldY, PATHING_TYPE_WALKABILITY))
            {
                newX = oldX;
            }

            if (IsTerrainPathable(oldX, newY, PATHING_TYPE_WALKABILITY))
            {
                newY = oldY;
            }

            SetUnitX(hero, newX);
            SetUnitY(hero, newY);
        }

        if (fwd == 0)
        {
            if (this.lastUnitAnimation != "stand")
            {
                this.lastUnitAnimation = "stand";
                hero.SetAnimation(0); // 0 is stand for most units
            }
        }
        else
        {
            if (this.lastUnitAnimation != "walk")
            {
                this.lastUnitAnimation = "walk";
                hero.SetAnimation(6); // POTM is 6 for walk
            }
        }

        Blizzard.SetCameraTargetControllerNoZForPlayer(player, hero, 0, 0, true);
        Blizzard.SetCameraFieldForPlayer(player, CAMERA_FIELD_ANGLE_OF_ATTACK, 310, 0);
        Blizzard.SetCameraFieldForPlayer(player, CAMERA_FIELD_FIELD_OF_VIEW, 1500, 0);
        Blizzard.SetCameraFieldForPlayer(player, CAMERA_FIELD_ROTATION, GetUnitFacing(hero), 0);
        Blizzard.SetCameraFieldForPlayer(player, CAMERA_FIELD_ZOFFSET, 100, 0);
    }

    public void SetKeyDownState(string key, bool state)
    {
        if (keyDownState.ContainsKey(key))
        {
            keyDownState[key] = state;
        }
    }

    private void ResetCamera()
    {
        if (!player.IsLocal) return;
        ResetToGameCamera(0);
        SetCameraField(CAMERA_FIELD_TARGET_DISTANCE, 2400.0f, 0.0f);
    }
}

public static class FirstPersonCameraManager
{
    private static Dictionary<player, FirstPersonCamera> cameras = new Dictionary<player, FirstPersonCamera>();

    public static void Initialize()
    {
        foreach (var player in Globals.ALL_PLAYERS)
        {
            var hero = GetHeroForPlayer(player);
            if (hero != null)
            {
                cameras[player] = new FirstPersonCamera(hero, player);
            }
        }

        RegisterKeyEvents();
    }

    private static void RegisterKeyEvents()
    {
        var keyStates = new Dictionary<bool, int>
    {
        { true, Blizzard.bj_KEYEVENTTYPE_DEPRESS },
        { false, Blizzard.bj_KEYEVENTTYPE_RELEASE }
    };

        var keys = new Dictionary<string, int>
    {
        { "UP", Blizzard.bj_KEYEVENTKEY_UP },
        { "DOWN", Blizzard.bj_KEYEVENTKEY_DOWN },
        { "LEFT", Blizzard.bj_KEYEVENTKEY_LEFT },
        { "RIGHT", Blizzard.bj_KEYEVENTKEY_RIGHT }
    };

        foreach (var keyState in keyStates)
        {
            foreach (var key in keys)
            {
                foreach (var p in Globals.ALL_PLAYERS)
                {
                    var keyTrigger = CreateTrigger();
                    var localKey = key.Key; // Create a local copy of the key
                    var localValue = keyState.Key; // Create a local copy of the value
                    Blizzard.TriggerRegisterPlayerKeyEventBJ(keyTrigger, p, keyState.Value, key.Value);
                    TriggerAddAction(keyTrigger, () => OnKeyEvent(localKey, localValue));
                }
            }
        }
    }

    private static void OnKeyEvent(string key, bool isDown)
    {
        var player = GetTriggerPlayer();
        if (cameras.ContainsKey(player))
        {
            cameras[player].SetKeyDownState(key, isDown);
        }
    }

    private static unit GetHeroForPlayer(player player)
    {
        return Globals.ALL_KITTIES.ContainsKey(player) ? Globals.ALL_KITTIES[player].Unit : null;
    }

    public static void ToggleFirstPerson(player player)
    {
        if (cameras.ContainsKey(player))
        {
            cameras[player].ToggleFirstPerson(!cameras[player].IsFirstPerson());
        }
    }

    public static void SetFirstPerson(player player, bool active)
    {
        if (cameras.ContainsKey(player))
        {
            cameras[player].ToggleFirstPerson(active);
        }
    }
}

File: src/Misc/FirstPersonCameraManager.ts
export class FirstPersonCamera {
    private timerPeriod: number = 0.001
    private FIRSTPERSON_ANGLE_PER_PERIOD: number = 0.3

    private forceCamTimer: Timer
    private hero: Unit
    private player: MapPlayer
    private keyDownState: Map<string, boolean>
    private lastUnitAnimation: string

    public FirstPersonCamera(hero: Unit, player: MapPlayer) {
        this.hero = hero
        this.player = player
        this.keyDownState = {
            UP: false,
            DOWN: false,
            LEFT: false,
            RIGHT: false,
        }
    }

    public IsFirstPerson(): boolean {
        return this.forceCamTimer != null
    }

    public ToggleFirstPerson(active: boolean) {
        if (active) {
            if (this.forceCamTimer == null) {
                this.forceCamTimer = Timer.create()
                TimerStart(this.forceCamTimer, this.timerPeriod, true, ErrorHandler.Wrap(UpdateCamera))
            }
        } else {
            if (this.forceCamTimer != null) {
                this.forceCamTimer.pause()
                DestroyTimer(this.forceCamTimer)
                this.ResetCamera()
            }
        }
    }

    private UpdateCamera() {
        if (this.hero == null || !this.hero.Alive) return

        let fwd: number = 0

        if (
            !(this.keyDownState['LEFT'] && this.keyDownState['RIGHT']) &&
            (this.keyDownState['LEFT'] || this.keyDownState['RIGHT'])
        ) {
            let angle: number = GetUnitFacing(this.hero)
            if (this.keyDownState['LEFT']) angle += this.FIRSTPERSON_ANGLE_PER_PERIOD
            if (this.keyDownState['RIGHT']) angle -= this.FIRSTPERSON_ANGLE_PER_PERIOD
            BlzSetUnitFacingEx(this.hero, angle)
        }

        if (
            !(this.keyDownState['UP'] && this.keyDownState['DOWN']) &&
            (this.keyDownState['UP'] || this.keyDownState['DOWN'])
        ) {
            let moveSpeed: number = GetUnitMoveSpeed(this.hero)
            let movePerTick: number = moveSpeed * timerPeriod

            let kitty: Kitty = Globals.ALL_KITTIES.get(player)

            if (kitty.Slider.IsEnabled()) {
                movePerTick = 0.2
            }

            let angle: number = Rad2Deg(GetUnitFacing(this.hero))
            if (this.keyDownState['UP']) fwd += movePerTick
            if (this.keyDownState['DOWN']) fwd -= movePerTick

            let oldX: number = this.hero.x
            let oldY: number = this.hero.y

            let newX: number = oldX + fwd * Cos(angle)
            let newY: number = oldY + fwd * Sin(angle)

            if (!Globals.GAME_ACTIVE && Regions.BarrierRegion.includes(newX, newY)) {
                return
            }

            if (IsTerrainPathable(newX, oldY, PATHING_TYPE_WALKABILITY)) {
                newX = oldX
            }

            if (IsTerrainPathable(oldX, newY, PATHING_TYPE_WALKABILITY)) {
                newY = oldY
            }

            this.hero.SetPathing(false)
            SetUnitPosition(this.hero, newX, newY)
            this.hero.SetPathing(true)
        }

        if (fwd == 0) {
            if (this.lastUnitAnimation != 'stand') {
                this.lastUnitAnimation = 'stand'
                hero.SetAnimation(0) // 0 is stand for most units
            }
        } else {
            if (this.lastUnitAnimation != 'walk') {
                this.lastUnitAnimation = 'walk'
                hero.SetAnimation(6) // POTM is 6 for walk
            }
        }

        SetCameraTargetControllerNoZForPlayer(player, hero, 0, 0, true)
        SetCameraFieldForPlayer(player, CAMERA_FIELD_ANGLE_OF_ATTACK, 310, 0)
        SetCameraFieldForPlayer(player, CAMERA_FIELD_FIELD_OF_VIEW, 1500, 0)
        SetCameraFieldForPlayer(player, CAMERA_FIELD_ROTATION, GetUnitFacing(hero), 0)
        SetCameraFieldForPlayer(player, CAMERA_FIELD_ZOFFSET, 100, 0)

        ItemPickup()
    }

    public SetKeyDownState(key: string, state: boolean) {
        if (keyDownState.has(key)) {
            keyDownState[key] = state
        }
    }

    private ResetCamera() {
        if (!player.isLocal()) return
        SetCameraTargetControllerNoZForPlayer(player, this.hero, 0, 0, false)
        ResetToGameCamera(0)
        SetCameraField(CAMERA_FIELD_TARGET_DISTANCE, 2400.0, 0.0)
    }

    private ItemPickup() {
        let kitty: Kitty = Globals.ALL_KITTIES.get(player)
        ItemSpatialGrid.KittyItemPickup(kitty)
    }
}

export class FirstPersonCameraManager {
    private static cameras: Map<player, FirstPersonCamera> = new Map()

    public static Initialize() {
        for (let player in Globals.ALL_PLAYERS) {
            let hero = GetHeroForPlayer(player)
            if (hero != null) {
                cameras[player] = new FirstPersonCamera(hero, player)
            }
        }

        RegisterKeyEvents()
    }

    private static RegisterKeyEvents() {
        let keyStates: Map<string, number> = {
            true: bj_KEYEVENTTYPE_DEPRESS,
            false: bj_KEYEVENTTYPE_RELEASE,
        }

        let keys: Map<string, number> = {
            UP: bj_KEYEVENTKEY_UP,
            DOWN: bj_KEYEVENTKEY_DOWN,
            LEFT: bj_KEYEVENTKEY_LEFT,
            RIGHT: bj_KEYEVENTKEY_RIGHT,
        }

        for (let keyState in keyStates) {
            for (let key in keys) {
                for (let p in Globals.ALL_PLAYERS) {
                    let keyTrigger = CreateTrigger()
                    let localKey = key.Key // Create a local copy of the key
                    let localValue = keyState.Key // Create a local copy of the value
                    TriggerRegisterPlayerKeyEventBJ(keyTrigger, p, keyState.Value, key.Value)
                    TriggerAddAction(keyTrigger, () => OnKeyEvent(localKey, localValue))
                }
            }
        }
    }

    private static OnKeyEvent(key: string, isDown: boolean) {
        if (cameras.has(GetTriggerPlayer())) {
            cameras[GetTriggerPlayer()].SetKeyDownState(key, isDown)
        }
    }

    private static GetHeroForPlayer(player: MapPlayer): Unit {
        return Globals.ALL_KITTIES.has(player) ? Globals.ALL_KITTIES.get(player).Unit : null
    }

    public static ToggleFirstPerson(player: MapPlayer) {
        if (cameras.has(player)) {
            cameras[player].ToggleFirstPerson(!cameras[player].IsFirstPerson())
        }
    }

    public static SetFirstPerson(player: MapPlayer, active: boolean) {
        if (cameras.has(player)) {
            cameras[player].ToggleFirstPerson(active)
        }
    }
}
